samples/nrf5340/empty_app_core/src/main.c

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright (c) 2020 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr.h>
+#include <init.h>
+#include <nrf.h>
+#include <nrfx.h>
+
+/** @brief Power OFF entire RAM and suspend CPU forever.
+ *
+ * Function operates only on `register` variables, so GCC will not use
+ * RAM after function prologue. This is also true even with optimizations
+ * disabled. Interrupts are disabled to make sure that they will never
+ * be executed when RAM is powered off.
+ *
+ * @param reg_begin  Address to `POWER` register of the first NRF_VMC->RAM item
+ * @param reg_last   Address to `POWER` register of the last NRF_VMC->RAM item
+ */
+void disable_ram_and_wfi(register volatile u32_t *reg_begin,
+			 register volatile u32_t *reg_last)
+{
+	__disable_irq();
+
+	do {
+		*reg_begin = 0;
+		reg_begin += sizeof(NRF_VMC->RAM[0]) / sizeof(reg_begin[0]);
+	} while (reg_begin <= reg_last);
+
+	do {
```

Comment:
you need barriers here, before WFI

Reply:
sort of nit: strictly speaking, you need to ensure BASEPRI is unlocking all interrupts , for WFI to be able to work as the spec intends it to work :)

Reply:
I think that I don't need a barrier:
```
#define __WFI()                             __ASM volatile ("wfi":::"memory")
```

I don't want any interrupt to occur, so BASEPRI should not unlock any interrupts. Or is it undefined behavior or invalid usage of WFI?

Reply:
I think that I don't need a barrier:
```
#define __WFI()                             __ASM volatile ("wfi":::"memory")
```

I don't want any interrupt to occur, so BASEPRI should not unlock any interrupts. Or is it undefined behavior or invalid usage of WFI?

Reply:
You need a data synchronization barrier, not a compiler synchronization barrier.

> I don't want any interrupt to occur, so BASEPRI should not unlock any interrupts. Or is it undefined behavior or invalid usage of WFI?

Interrupts won't occur because you are setting PRIMASK to 1.
BASEPRI should be unconditionally unlocked, to allow the while loop to work properly. It's more of a nit, but will help will possible debug issues.

Reply:
@doki-nordic for your reference you can consult how the ARCHitecture code implements the CPU idling mechanism

Reply:
Yes, I missed that `dsb` should be added before `wfi`.

I still don't see the reason for setting BASEPRI. I agree that in normal CPU idling mechanism it is important, but this case not normal, because I want to stay in wfi instruction FOREVER. I don't need to:
> allow the while loop to work properly

Reply:
After some more reading I can see that setting BASEPRI to the highest possible value will be nice in this case which is the opposite of CPU idling mechanism, but is not not neccessary.

@ioannisg please remember that this loop is not CPU idling mechanism. Intention is to completely stop the CPU.

Reply:
you don't need the ISB, only DSB is needed. I think adding DSB outside the while(1) loop should be enough.

Reply:
Regarding BASEPRI, yeap, I said it's more of a nit. You can ignore it. No, setting it to highest is not needed since you have a while loop anyways.

Reply:
I added barriers before I've read comment above. I added both ISB and DSB, because I saw that some implementation also use ISB before WFI. I put it into loop. Is it ok for you @ioannisg?

Reply:
Can you do the following:

```
        __DSB();
	do {
		__WFI();
	} while (1);
```

Summarizing reasoning:
- ISB is not required before WFE/WFI. There's no need to flush the instruction pipeline before WFI, WFE.
- DSB is required before WFE, WFI because you need to make sure that all memory accesses are completed. This is a generic requirement. It might not be required given the particular use case, but it's no harm if you add it (then you don't need to care at all). But add it once, before the loop only: inside the loop there is no need to have it since you do not perform memory accesses.

Reply:
Done